qemu/tests/qemu_guest_agent.py

```diff
@@ -897,6 +900,81 @@ def gagent_check_thaw_unfrozen(self, test, params, env):
             test.fail("The return value of thawing an unfrozen fs is %s,"
                       "it should be zero" % ret)
 
+    @error_context.context_aware
+    def gagent_check_hotplug_frozen(self, test, params, env):
+        """
+        hotplug device with frozen fs
+
+        :param test: kvm test object
+        :param params: Dictionary with the test parameters
+        :param env: Dictionary with test environment
+        """
+        def find_image(image_name):
+            """
+            Find the path of the iamge.
+            """
+            image_params = params.object_params(image_name)
+            o = storage.get_image_filename(image_params, data_dir.get_data_dir())
+            return o
+
+        def get_new_disk(disk1, disk2):
+            """
+            Get the different disk between disk1 and disk2.
+            """
+            disk = list(set(disk2.split("\n")).difference(set(disk1.split("\n"))))
+            return disk
+
+        session = self._get_session(params, self.vm)
+        disk_mount_point = params["disk_mount_point"]
+        if params.get("os_type") == "linux":
+            session.cmd("mkdir %s" % disk_mount_point)
+            get_disk_cmd = params["get_disk_cmd"]
+            disks_before_plug = session.cmd_output_safe(get_disk_cmd)
+        error_context.context("Freeze guest fs", logging.info)
+        self.gagent.fsfreeze()
+        error_context.context("Hotplug a disk to guest", logging.info)
+        pci_type = params["pci_type"]
+        images = params["images"].split()
+        controller_model = params["controller_model"]
+        disk_write_cmd = params["disk_write_cmd"]
+        img_format_type = params["img_format_type_stg0"]
+        pause = float(params.get("virtio_block_pause", 10.0))
+        device = qdevices.QDevice(pci_type)
+        controller = qdevices.QDevice(controller_model, params={"id": "scsi1"})
+        controller.hotplug(self.vm.monitor)
+        drive = qdevices.QRHDrive("block1")
+        drive.set_param("file", find_image(images[1]))
+        drive.set_param("format", img_format_type)
+        drive_id = drive.get_param("id")
+        drive.hotplug(self.vm.monitor)
+        device.set_param("drive", drive_id)
+        device.set_param("id", "scsi-disk1")
+        device.hotplug(self.vm.monitor)
+        error_context.context("Format and write disk", logging.info)
+        if params.get("os_type") == "linux":
+            plug_status = utils_misc.wait_for(lambda: len(get_new_disk(disks_before_plug,
+                                              session.cmd_output_safe(get_disk_cmd))) != 0, pause)
+            if plug_status:
+                disks_after_plug = session.cmd_output_safe(get_disk_cmd)
+                new_disks = get_new_disk(disks_before_plug, disks_after_plug)
+            else:
+                test.fail("Can't get new disks")
+            session.cmd("mkfs.xfs %s && mount %s %s" %
```

Comment:
> if guestFS is frozen. mkfs will fail. so should check guest status first.

If we execute mkfs after frozen by auto, it will pass.  Because mkfs cann't be executed after several seconds of frozen, not immediately.